src/elm/interval.coffee

```diff
@@ -345,27 +344,71 @@ module.exports.Expand = class Expand extends Expression
     # return null if unit conversion failed, must have mismatched units
     return null if not (low_value? and high_value? and per_value?)
 
-    results = @makeNumericIntervalList(
-      low_value, high_value, interval.lowClosed, interval.highClosed, per_value)
+    results = @makeNumericIntervalList(low_value, high_value, interval.lowClosed, interval.highClosed, per_value)
 
     for itvl in results
       itvl.low = new Quantity(itvl.low, result_units)
       itvl.high = new Quantity(itvl.high, result_units)
     return results
 
   expandNumericInterval: (interval, per) ->
-    return null if per.unit != '1'
-    return @makeNumericIntervalList(
-      interval.low, interval.high, interval.lowClosed, interval.highClosed, per.value)
-
-  makeNumericIntervalList: (low, high, lowClosed, highClosed, per) ->
-    point_intervals = Number.isInteger(low) and Number.isInteger(high) and Number.isInteger(per)
-    gap = if point_intervals then 1 else MIN_FLOAT_PRECISION_VALUE
-    low = low + gap if !lowClosed
-    high = high - gap if !highClosed
-    width = per - gap
+    return null unless per.unit is '1' or per.unit is ''
+    @makeNumericIntervalList(interval.low, interval.high, interval.lowClosed, interval.highClosed, per.value)
+
+  makeNumericIntervalList: (low, high, lowClosed, highClosed, perValue) ->
+    # Determine the precision to pass to the truncateDecimal call below.
+    if Math.floor(perValue.valueOf()) == perValue.valueOf()
+      decimalPrecision = 0
+    else
+      decimalPrecision = perValue.toString().split('.')[1].length || 0
+
+    # CQL has max decimal precision of 8
+    if decimalPrecision > 8
+      decimalPrecision = 8
+
+    low = if lowClosed then low else successor low
+    high = if highClosed then high else predecessor high
+
+    # If the interval boundaries are more precise than the per quantity, the
+    # more precise values will be truncated to the precision specified by the
+    # per quantity.
+    low = truncateDecimal(low, decimalPrecision)
+    high = truncateDecimal(high, decimalPrecision)
+
     return [] if low > high
-    results = (new dtivl.Interval(x, x + width, true, true) for x in [low..(high - width)] by per)
+    return [] unless low? and high?
+
+    # Deal with the weird case of a point interval expanding with more
+    # precision than it is defined by moving the high
+    if low is high
```

Comment:
Check if potentially integers here

No way to tell if only 0s after the decimal, but can if there are nonzero decimal digits.